auth/src/main/java/com/redhat/lightblue/rest/auth/ldap/LdapFindUserByUidCommand.java

```diff
@@ -1,62 +1,122 @@
     }
 
     @Override
     protected SearchResult run() throws Exception {
-        String searchFilter = "(uid=" + uid + ")";
+        SearchResult searchResult = null;
 
-        SearchControls searchControls = new SearchControls();
-        searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+        try {
+            searchResult = LDAPSearcher.searchLDAPServer(cacheKey);
+        } catch (NamingException e) {
+           //propagate the exception
+            throw e;
+        } catch (LDAPUserNotFoundException | LDAPMutipleUserFoundException e) {
+            searchResult = LDAPCache.getLDAPCacheSession().getIfPresent(cacheKey);
+            // if (not found on the server OR server state is inconsistent ) and cache hold the old value, evict the entry
+            if (searchResult != null) {
+                // invalidate all caches
+                LDAPCache.invalidateKey(cacheKey);
+            }
+        }
+        LDAPCache.getLDAPCacheSession().put(cacheKey, searchResult);
 
-        NamingEnumeration<SearchResult> results = ldapContext.search(ldapSearchBase, searchFilter, searchControls);
+        return searchResult;
+    }
 
-        SearchResult searchResult = null;
-        if (results.hasMoreElements()) {
-            searchResult = results.nextElement();
+    /*
+        This methods is executed for all types of failure such as run() failure, timeout, thread pool or semaphore rejection, and circuit-breaker short-circuiting
+     */
+    @Override
+    protected SearchResult getFallback() {
+        LOGGER.info("Error during the execution of the command. Falling back to the cache");
+        return new FallbackViaLDAPServerProblemCommand(cacheKey, getFailedExecutionException()).execute();
 
-            //make sure there is not another item available, there should be only 1 match
-            if (results.hasMoreElements()) {
-                LOGGER.error("Matched multiple users for the accountName: " + uid);
+    }
+
+
+    /*
+        Use the cache in case tje LDAP Server was not available and also to we have metrics around the fallback
+     */
+    private static class FallbackViaLDAPServerProblemCommand extends HystrixCommand<SearchResult> {
+        private static final Logger LOGGER = Logger.getLogger(FallbackViaLDAPServerProblemCommand.class);
+
+        private final LDAPCacheKey cacheKey;
+        private final Throwable failedExecutionException;
+
+        public FallbackViaLDAPServerProblemCommand(LDAPCacheKey cacheKey, Throwable failedExecutionException) {
+            super(HystrixCommand.Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey(GROUPKEY)).
+                    andCommandKey(HystrixCommandKey.Factory.asKey(GROUPKEY + ":" + FallbackViaLDAPServerProblemCommand.class.getSimpleName())));
+            this.cacheKey = cacheKey;
+            this.failedExecutionException = failedExecutionException;
+        }
+
+        @Override
+        protected SearchResult run() throws Exception {
```

Comment:
OK. I will remove the retry from cache ( I sort of thought that would be nice to retry but might need a more careful plan how we would do that)